docs/smart-contracts/guides/conventions/work-contractspec-js.mdx

```diff
@@ -4,4 +4,696 @@ hide_table_of_contents: true
+   ```
+
+5. **Convert to JSON Schema**: The `jsonSchema` method converts the contract specification to a JSON schema, which can be useful for documentation or generating client code.
+
+   ```typescript
+   const schema = contractSpec.jsonSchema();
+   const funcSchema = contractSpec.jsonSchema("funcName");
+   ```
+
+By leveraging these methods, you can easily convert between native JavaScript values and the `ScVal` format required for interacting with Soroban smart contracts, handle errors effectively, and generate useful documentation or client code based on the contract specification.
+
+#### Step-by-Step Guide
+
+##### Example Contract: TokenManager
+
+Here's an example Rust contract that we'll be working with. This contract, `TokenManager`, manages token balances and includes functions to initialize, deposit, withdraw, and check balances. It also includes custom errors for different failure scenarios.
+
+```rust
+use soroban_sdk::{contract, contractimpl, Address, Env, Symbol, contracterror};
+
+#[contract]
+pub struct TokenManager;
+
+#[contracterror]
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum TokenManagerError {
+    InternalError = 1,
+    UnauthorizedError = 2,
+    InsufficientBalanceError = 3,
+    OverflowError = 4,
+}
+
+#[contractimpl]
+impl TokenManager {
+    pub fn initialize(e: Env, admin: Address) {
+        e.storage().set(&Symbol::short("admin"), &admin);
+    }
+
+    pub fn deposit(e: Env, user: Address, amount: i128) -> Result<(), TokenManagerError> {
+        let key = Symbol::short("balance").join(&user);
+        let mut balance: i128 = e.storage().get(&key).unwrap_or(0);
+        balance = balance.checked_add(amount).ok_or(TokenManagerError::OverflowError)?;
+        e.storage().set(&key, &balance);
+        Ok(())
+    }
+
+    pub fn withdraw(e: Env, user: Address, amount: i128) -> Result<(), TokenManagerError> {
+        let key = Symbol::short("balance").join(&user);
+        let mut balance: i128 = e.storage().get(&key).unwrap_or(0);
+        if balance < amount {
+            return Err(TokenManagerError::InsufficientBalanceError);
+        }
+        balance -= amount;
+        e.storage().set(&key, &balance);
+        Ok(())
+    }
+
+    pub fn get_balance(e: Env, user: Address) -> i128 {
+        let key = Symbol::short("balance").join(&user);
+        e.storage().get(&key).unwrap_or(0)
+    }
+}
+```
+
+There are a few options to get your contract Spec object to work with in javascript. The Spec is constructed from a (xdr.ScSpecEntry)[https://github.com/stellar/js-stellar-base/blob/5efdff0b2a78ed842f7463340b4ff23eb4370635/xdr/curr/Stellar-contract-spec.x#L228]
+
+#### Option 1: Create Contract Specification in JavaScript
+
+To create a contract specification from scratch in JavaScript, we need to define the functions and errors of the contract. Here's how you can do it:
+
+```typescript
+import { xdr } from "@stellar/stellar-base";
+import { Spec } from "@stellar/stellar-sdk";
```

Comment:
`Spec` needs to be imported from the `contract` entrypoint:
```suggestion
import { xdr } from "@stellar/stellar-sdk";
import { Spec } from "@stellar/stellar-sdk/contract";
```

...or used from the `contract` import:
```suggestion
import { xdr, contract } from "@stellar/stellar-sdk";
// later, can use `contract.Spec` instead of just `Spec`
```